### PR TITLE
Implement BLS backend

### PR DESCRIPTION
**Pull request by krishung5**
Previous PR: https://github.com/triton-inference-server/bls_backend/pull/1
Added L0_backend_bls test: https://github.com/triton-inference-server/server/pull/4494

### CODE REVIEW

File: examples/README.md
Comment (GuanLuo) [inline examples/README.md:58]: Actually, I wonder if we need to put the BLS example in a separate GitHub repository. For example, in the below section, we have some backend examples that are placed [here](https://github.com/triton-inference-server/backend/tree/main/examples). We can put BLS example in the same directory and add the pointer here. CC @tanmayv25

Comment (krishung5) [reply examples/README.md:58]: I'm fine with either way, but to me it looks like minimal and recommended backends are some preliminary examples for creating user's own backend and are covered in the [Tutorial](https://github.com/triton-inference-server/backend/tree/main/examples#tutorial) whereas repeat and stateful backends demonstrate some advanced usage and are not covered in the Tutorial. I think it would be better if @tanmayv25 can make the final decision here.

Comment (krishung5) [reply examples/README.md:58]: Moving to "backend/examples/backends/" makes more sense now. thank you both for the suggestion. With the PR under bls_backend repo is under review, should I make this change after the PR is approved?

Comment (GuanLuo) [reply examples/README.md:58]: I am currently reviewing the PR in `bls_backend` repository, I think you can keep the existing PR for now to collect the 2nd round of review comment, and have a separate PR in `backend` repo with the up-to-date code change and start a new round of review in the new PR.

File: examples/backends/bls/src/bls_utils.cc
Comment (Tabrizian) [inline examples/backends/bls/src/bls_utils.cc:234]: I think we don't need to create a new response for every request. This can be an internal field created in the constructor.

File: examples/backends/bls/src/bls_utils.cc
Comment (Tabrizian) [inline examples/backends/bls/src/bls_utils.cc:255]: This method should not be part of the BLSExecutor class.

File: examples/backends/bls/src/bls_utils.cc
Comment (Tabrizian) [inline examples/backends/bls/src/bls_utils.cc:223]: Fail to execute the inference if the model is using the decoupled transaction policy. You can check the model's transaction policy for this: https://github.com/triton-inference-server/python_backend/blob/3a60cfc7fb3525d1200ee179a1355fd813cccd28/src/request_executor.cc#L189-L199

File: examples/backends/bls/src/bls_utils.cc
Comment (Tabrizian) [inline examples/backends/bls/src/bls_utils.cc:119]: Add a comment block explaining that the following logic only works for non-decoupled models and needs to be modified if the model is using decoupled API.

File: examples/backends/bls/src/bls.cc
Comment (Tabrizian) [inline examples/backends/bls/src/bls.cc:369]: There is memory leak because bls_executor is never deleted. You can just use a local variable.

File: examples/backends/bls/src/bls.cc
Comment (Tabrizian) [inline examples/backends/bls/src/bls.cc:323]: Can this check and the check below it be part of the BLSExecutor `Execute` function?

File: examples/backends/bls/src/bls.cc
Comment (Tabrizian) [inline examples/backends/bls/src/bls.cc:375]: I'm not fully clear on why this variable is required. If we are in the catch block, `is_inference_success` will always be false.

File: examples/backends/bls/src/bls_utils.cc
Comment (Tabrizian) [inline examples/backends/bls/src/bls_utils.cc:86]: Use LOG_MESSAGE instead of std_cout.

File: examples/backends/bls/src/bls_utils.cc
Comment (Tabrizian) [inline examples/backends/bls/src/bls_utils.cc:52]: Same

File: examples/backends/bls/src/bls_utils.cc
Comment (Tabrizian) [inline examples/backends/bls/src/bls_utils.cc:64]: Same

File: examples/backends/bls/src/bls_utils.cc
Comment (Tabrizian) [inline examples/backends/bls/src/bls_utils.cc:94]: Use LOG_ERROR here

File: examples/backends/bls/src/bls_utils.cc
Comment (Tabrizian) [inline examples/backends/bls/src/bls_utils.cc:167]: Why not pass the pointer here instead of double pointer?

File: examples/backends/bls/src/bls_utils.cc
Comment (Tabrizian) [inline examples/backends/bls/src/bls_utils.cc:206]: Same

File: examples/backends/bls/src/bls.cc
Comment (krishung5) [reply examples/backends/bls/src/bls.cc:323]: Does it make sense if we make sure both models are ready before starting to send two requests on them? There might be a scenario that the first model is ready and running inference while the second model is not ready and throw and exception here. With the exception, we will move on to process the next request while the status of the first model is hard to handle

File: examples/backends/bls/src/bls.cc
Comment (krishung5) [reply examples/backends/bls/src/bls.cc]: Thanks for the catch. Removed `is_inference_success `.

**Review by Tabrizian — COMMENTED** [review]
I think having both BLSExecutor and ModelExecutor is confusing for this example. Let's stick to only using ModelExecutor to keep this example simple. The ModelExecutor should be very similar to what we had in the previous commit with just the comments addressed. Also, now it looks like none of the BLSExecutor and ModelExecutor classes have the abstraction we previously had in "BLSExecutor" as they both kind of depend on the specific pipeline we have here.

**Comment by krishung5** [comment]
> I think having both BLSExecutor and ModelExecutor is confusing for this example. Let's stick to only using ModelExecutor to keep this example simple. The ModelExecutor should be very similar to what we had in the previous commit with just the comments addressed. Also, now it looks like none of the BLSExecutor and ModelExecutor classes have the abstraction we previously had in "BLSExecutor" as they both kind of depend on the specific pipeline we have here.

I might be wrong here but I was under the impression from this [comment](https://github.com/triton-inference-server/bls_backend/pull/1#discussion_r899572621) that we would like to have a `BLSExecutor` that encapsulates the whole pipeline. Users can put their own custom logic in `BLSExecutor`, while the `ModelExecutor` object only handles each single internal inference. Not sure if renaming `ModelExecutor` to `RequestExecutor` and adding more detailed comments would help to make it less confused. If not, I can revert the changes back.

File: examples/backends/bls/src/bls_utils.h
Comment (Tabrizian) [inline examples/backends/bls/src/bls_utils.h:62]: Move the allocator to this class since BLS executor doesn't need it. `TRITONSERVER_Server* server` should be an internal field of `ModelExecutor`.

File: examples/backends/bls/src/bls_utils.h
Comment (Tabrizian) [inline examples/backends/bls/src/bls_utils.h:66]: We would no longer need these two arguments after making the change above.

Comment (Tabrizian) [inline examples/backends/bls/src/bls_utils.h:66]: We would no longer need these two arguments after making the change above.

File: examples/backends/bls/src/bls_utils.cc
Comment (Tabrizian) [inline examples/backends/bls/src/bls_utils.cc:185]: Make the ModelExecutor an internal field of BLSExecutor so that we don't have to create a new ModelExecutor for each inference request.

Comment (Tabrizian) [inline examples/backends/bls/src/bls_utils.cc:185]: Make the ModelExecutor an internal field of BLSExecutor so that we don't have to create a new ModelExecutor for each inference request.

Comment (Tabrizian) [inline examples/backends/bls/src/bls_utils.cc:185]: The allocator will be part of the Model Executor.

Comment (Tabrizian) [inline examples/backends/bls/src/bls_utils.cc:185]: The allocator will be part of the Model Executor.

File: examples/backends/bls/src/bls_utils.h
Comment (krishung5) [reply examples/backends/bls/src/bls_utils.h]: Thanks for the catch. Moved allocator and `TRITONSERVER_Server* server` to `ModelExecutor` as internal fields.

File: examples/backends/bls/src/bls_utils.cc
Comment (Tabrizian) [inline examples/backends/bls/src/bls_utils.cc:336]: Move this class to the constructor. Also, no need to use a unique_ptr. You can just store it in a regular local variable.

**Review by Tabrizian — APPROVED** [review]
One minor comment. Otherwise, LGTM.

File: examples/backends/bls/src/bls.cc
Comment (Tabrizian) [reply examples/backends/bls/src/bls.cc]: It is fine to have it here for now. But we would need to include them in the ModelExecutor for the final BLS utility.

**Comment by GuanLuo** [comment]
> I think having both BLSExecutor and ModelExecutor is confusing for this example

> Update: Synced with Iman offline. Moving PrepareInferenceRequest, PrepareInferenceInput and PrepareInferenceOutput functions to BLSExecutor class would be more accurate, making ModelExecutor not depend on the specific BLS logic we have here.

I think we should reorganize the files to be even more clear. Again, when users look at this example, they are more interested in how to write their BLS logic (`BLSExecutor`), we can have:
- `backend.cc`: renamed from the `bls.cc`, the content of the file is to implement the required Triton backend functions, this is not BLS specific. In this file, we should point out that this is standard for any backend implementation and how the BLS logic eventually get invoked (in `TRITONBACKEND_ModelInstanceExecute`).
- `bls.cc`: where the BLS (`BLSExecutor`) of this example locates, user should refer to this file for how to interact with Triton in-process API to build the execution pipeline.
- `bls_utils.cc`: as the name suggested, where the utilities that are not BLS dependent (`ModelExecutor` / common allocators) locate. This will be the first draft of the BLS utilities that the user can take advantage of, but I think it should be kept as part of the example for now, and we can extract it to the actual utility directory once we are confident with the quality.

**Comment by krishung5** [comment]
> > I think having both BLSExecutor and ModelExecutor is confusing for this example
> 
> > Update: Synced with Iman offline. Moving PrepareInferenceRequest, PrepareInferenceInput and PrepareInferenceOutput functions to BLSExecutor class would be more accurate, making ModelExecutor not depend on the specific BLS logic we have here.
> 
> I think we should reorganize the files to be even more clear. Again, when users look at this example, they are more interested in how to write their BLS logic (`BLSExecutor`), we can have:
> 
> * `backend.cc`: renamed from the `bls.cc`, the content of the file is to implement the required Triton backend functions, this is not BLS specific. In this file, we should point out that this is standard for any backend implementation and how the BLS logic eventually get invoked (in `TRITONBACKEND_ModelInstanceExecute`).
> * `bls.cc`: where the BLS (`BLSExecutor`) of this example locates, user should refer to this file for how to interact with Triton in-process API to build the execution pipeline.
> * `bls_utils.cc`: as the name suggested, where the utilities that are not BLS dependent (`ModelExecutor` / common allocators) locate. This will be the first draft of the BLS utilities that the user can take advantage of, but I think it should be kept as part of the example for now, and we can extract it to the actual utility directory once we are confident with the quality.

Thanks @GuanLuo. I think reorganizing the files as you mentioned does make it a clearer example. I will reorganize the files and add the description about how users should refer to these files to the document.

File: examples/backends/bls/src/bls_utils.cc
Comment (krishung5) [reply examples/backends/bls/src/bls_utils.cc]: Updated, please review.

File: examples/backends/bls/src/bls.cc
Comment (krishung5) [reply examples/backends/bls/src/bls.cc]: I will keep this in mind for including these check in the `ModelExecutor `.

**Review by GuanLuo — APPROVED** [review]
CC @tanmayv25

**Review by Tabrizian — APPROVED** [review]

File: examples/README.md
Comment (tanmayv25) [inline examples/README.md:431]: Add a section here talking about the self-imposed limitations that were made for the simplicity of example.
Some of them that I see are: no batching, no decoupled models, no GPU tensors etc.

There might be some that I missed.

File: examples/backends/bls/src/backend.cc
Comment (tanmayv25) [inline examples/backends/bls/src/backend.cc:38]: Remove L38. 
It is little confusing to think that BLS backend supports a model. When it is actually a backend connecting multiple models in itself.

**Review by tanmayv25 — CHANGES_REQUESTED** [review]

File: examples/backends/bls/src/bls.cc
Comment (krishung5) [reply examples/backends/bls/src/bls.cc]: Thanks for the heads up. Will add the description about not supporting decoupled when writing backend utilities.

File: examples/README.md
Comment (krishung5) [reply examples/README.md:431]: Added the limitation of model configuration setting.

**Pull request merged**
